src/services/route.rs

```diff
@@ -0,0 +1,257 @@
+    /// An error indicating that the string passed to the `RouteInfo::parse()` function did not contain ary characters
+    RouteIsEmpty,
+    /// Indicates that the url could not be retrieved from the Location API.
+    CouldNotGetLocationHref
+}
+
+
+impl RouteInfo {
+    /// This expects a string with a leading slash`
+    pub fn parse(route_string: &str) -> Result<RouteInfo, RoutingError> {
+        // Perform some validation on the string before parsing it.
+        if let Some(first_character) = route_string.chars().next() {
+            if first_character != '/' {
+                eprintln!("does not start with slash: '{}'", route_string);
+                return Err(RoutingError::RouteDoesNotStartWithSlash)
+            }
+        } else {
+            return Err(RoutingError::RouteIsEmpty)
+        }
+
+        let full_url = format!("http://dummy_url.com{}", route_string);
+        Url::parse(&full_url)
+            .map(RouteInfo::from)
+            .map_err(|_| RoutingError::CouldNotParseRoute { route: route_string.to_string() })
+    }
+
+    /// Converts the RouteInfo into a string that can be matched upon,
+    /// as well as stored in the History Api.
+    pub fn to_string(&self) -> String {
+        let path = self.path_segments.join("/");
+        let mut path = format!("/{}", path); // add the leading '/'
+        if let Some(ref query) = self.query {
+            path = format!("{}?{}", path, query);
+        }
+        if let Some(ref fragment) = self.fragment {
+            path = format!("{}#{}", path, fragment)
+        }
+        path
+    }
+
+    /// Gets the path segment at the specified index.
+    pub fn get_segment_at_index<'a>(&'a self, index: usize) -> Option<&'a str> {
+        self.path_segments.get(index).map(String::as_str)
+    }
+}
+
+impl From<Url> for RouteInfo {
+    fn from(url: Url) -> RouteInfo {
+        RouteInfo {
+            path_segments: url
+                .path_segments()
+                .expect("The route should always start with '/', so this should never error.")
+                .map(str::to_string)
+                .collect::<Vec<String>>(),
+            query: url.query().map(str::to_string),
+            fragment: url.fragment().map(str::to_string)
+        }
+    }
+}
+
+
+impl RouteService {
+
+    /// Creates a new route service
+    pub fn new() -> RouteService {
+        RouteService {
+            history: window().history(),
+            location: window().location().unwrap(),
+            event_listener: None,
+            callback: None
+        }
+    }
+
+
+    /// Creates the callback used in the main routing logic.
+    ///
+    /// The callback takes a string, parses it into a url, and then uses the result of that
+    /// to create a message that the component will use to update itself with.
+    // TODO I would like to include this in register_router
+    pub fn create_routing_callback<COMP, CTX>(context: &mut Env<CTX, COMP>) -> Callback<RouteResult>
```

Comment:
I think this should be removed. We are just calling `context.send_back`, which is required for _all_ services. I don't think this adds anything.

Reply:
I also disagree with the API of requiring the user to implement `COMP::Msg::from(route_result)`. The callback can handle that by itself if we used it correctly: the user created the `Callback` using a function of type `Fn(RouteInfo) -> Msg`

Reply:
I'm inclined to agree. I wanted the code within `create_routing_callback()` to live within `register_router()` but because I couldn't both mutably borrow the context to get access to the router to set call that function, as well as pass a mutable reference to it as a parameter to that same function, I ended up creating this awkward function that needs to be called beforehand.

Because everything else requires calling `context.send_back()`, I don't see the problem with this.